File: app/models/note_model.py
"""
笔记类型（NoteModel）和卡片模板（CardTemplate）模型

NoteModel 定义字段结构，CardTemplate 定义问答模板
"""

from sqlalchemy import JSON, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, BaseTableMixin


class NoteModel(Base, BaseTableMixin):
    """笔记类型模型 - 定义笔记的字段结构"""

    __tablename__ = "note_models"

    user_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id"), nullable=False, index=True, comment="所属用户ID"
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False, comment="笔记类型名称")
    fields_schema: Mapped[dict] = mapped_column(
        JSON,
        nullable=False,
        default=list,
        comment="字段定义，如 [{'name': 'Front'}, {'name': 'Back'}]",
    )
    css: Mapped[str | None] = mapped_column(Text, nullable=True, comment="自定义CSS样式")

    # 关系
    templates: Mapped[list["CardTemplate"]] = relationship(
        "CardTemplate", back_populates="note_model", cascade="all, delete-orphan", lazy="selectin"
    )

    def __repr__(self) -> str:
        return f"<NoteModel(id={self.id}, name={self.name})>"


class CardTemplate(Base, BaseTableMixin):
    """卡片模板模型 - 定义问答侧的 HTML 模板"""

    __tablename__ = "card_templates"

    note_model_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("note_models.id"), nullable=False, index=True, comment="所属笔记类型ID"
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False, comment="模板名称")
    ord: Mapped[int] = mapped_column(Integer, nullable=False, default=0, comment="模板序号（从0开始）")
    question_template: Mapped[str] = mapped_column(Text, nullable=False, default="{{Front}}", comment="问题侧HTML模板")
    answer_template: Mapped[str] = mapped_column(
        Text, nullable=False, default="{{FrontSide}}<hr>{{Back}}", comment="答案侧HTML模板"
    )

    # 关系
    note_model: Mapped["NoteModel"] = relationship("NoteModel", back_populates="templates")

    def __repr__(self) -> str:
        return f"<CardTemplate(id={self.id}, name={self.name}, ord={self.ord})>"

File: web/src/db/repositories/noteModel.ts
/**
 * NoteModel 本地存储操作
 */

import { nanoid } from 'nanoid'
import { db } from '@/db'
import type { NoteModel, CardTemplate, FieldDefinition } from '@/types'

export const noteModelRepo = {
  /**
   * 获取所有笔记类型
   */
  async getAll(userId: string): Promise<NoteModel[]> {
    return db.noteModels
      .where('userId')
      .equals(userId)
      .and((nm) => !nm.deletedAt)
      .toArray()
  },

  /**
   * 根据 ID 获取笔记类型
   */
  async getById(id: string): Promise<NoteModel | undefined> {
    const noteModel = await db.noteModels.get(id)
    if (noteModel && !noteModel.deletedAt) {
      // 加载关联的模板
      const templates = await db.cardTemplates
        .where('noteModelId')
        .equals(id)
        .and((t) => !t.deletedAt)
        .sortBy('ord')
      noteModel.templates = templates
      return noteModel
    }
    return undefined
  },

  /**
   * 创建笔记类型
   */
  async create(
    userId: string,
    name: string,
    fieldsSchema: FieldDefinition[],
    templates: Omit<CardTemplate, 'id' | 'noteModelId' | 'createdAt' | 'updatedAt'>[],
    css?: string
  ): Promise<NoteModel> {
    const now = Date.now()
    const noteModelId = nanoid()

    const noteModel: NoteModel = {
      id: noteModelId,
      userId,
      name,
      fieldsSchema,
      css,
      templates: [],
      createdAt: now,
      updatedAt: now,
    }

    // 创建模板
    const cardTemplates: CardTemplate[] = templates.map((t, index) => ({
      id: nanoid(),
      noteModelId,
      name: t.name,
      ord: t.ord ?? index,
      questionTemplate: t.questionTemplate,
      answerTemplate: t.answerTemplate,
      createdAt: now,
      updatedAt: now,
    }))

    await db.transaction('rw', [db.noteModels, db.cardTemplates], async () => {
      await db.noteModels.add(noteModel)
      await db.cardTemplates.bulkAdd(cardTemplates)
    })

    noteModel.templates = cardTemplates
    return noteModel
  },

  /**
   * 更新笔记类型
   */
  async update(
    id: string,
    data: Partial<Pick<NoteModel, 'name' | 'fieldsSchema' | 'css'>>
  ): Promise<void> {
    await db.noteModels.update(id, {
      ...data,
      updatedAt: Date.now(),
    })
  },

  /**
   * 删除笔记类型（软删除）
   */
  async delete(id: string): Promise<void> {
    const now = Date.now()
    await db.transaction('rw', [db.noteModels, db.cardTemplates], async () => {
      await db.noteModels.update(id, { deletedAt: now, updatedAt: now })
      await db.cardTemplates
        .where('noteModelId')
        .equals(id)
        .modify({ deletedAt: now, updatedAt: now })
    })
  },

  /**
   * 创建默认的 Basic 笔记类型
   */
  async createDefaultBasic(userId: string): Promise<NoteModel> {
    return this.create(
      userId,
      'Basic',
      [{ name: 'Front' }, { name: 'Back' }],
      [
        {
          name: '正向',
          ord: 0,
          questionTemplate: '<div class="card-front">{{Front}}</div>',
          answerTemplate: '<div class="card-front">{{Front}}</div><hr><div class="card-back">{{Back}}</div>',
        },
      ],
      `.card-front { font-size: 1.5rem; text-align: center; }
.card-back { font-size: 1.2rem; text-align: center; color: #666; }`
    )
  },
}



